core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -636,6 +637,23 @@ public QueryBuilder readQuery() throws IOException {
         return readNamedWriteable(ScoreFunctionBuilder.class);
     }
 
+    @FunctionalInterface
+    public interface Reader<T> {
+        T read(StreamInput t) throws IOException;
+    }
+
+    /**
+     * Reads a list of objects
+     */
+    public <T> List<T> readList(Reader<T> producer) throws IOException {
+        List<T> builder = new ArrayList<>();
+        int count = readVInt();
```

Comment:
I suspect you can build the arraylist here with the expected size being the count.